### PR TITLE
feat: support sync mode full refresh

### PR DESCRIPTION
**Pull request by afthabvp**
## Description

#105 

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would impact existing functionality)
- [ ] Documentation update
- [ ] Chore
  
## Checklist:
- [x] Ensure a branch name is prefixed with `feature`, `bugfix`, `hotfix`, `release` or `chore` followed by `/` and branch name e.g `feature/add-salesforce-connector`
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/reverse_etl/extractors/full_refresh.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+        return log_sync_run_error(sync_run) unless sync_run.may_query?
+
+        sync_run.query!
+
+        source_client = setup_source_client(sync_run.sync)
+        model = sync_run.sync.model
+        batch_query_params = batch_params(source_client, sync_run)
+
+        prepare_for_insertion(sync_run)
```
Comment (subintp) [inline lib/reverse_etl/extractors/full_refresh.rb:18]: Name can be flush_records

Comment (subintp) [inline lib/reverse_etl/extractors/full_refresh.rb:18]: Why are we clearing the records in sync fashion here ? We can do this async via a temporal activity once after every full refresh sync in that way the any failures on deletes are not affected on the sync

File: lib/reverse_etl/extractors/full_refresh.rb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        return if records.count == sync_records_to_save.count && sync_records_to_save.count == result.rows.size
+
+        log_mismatch_error(records, sync_records_to_save, result.rows.flatten,
+                           sync_run)
+      rescue StandardError => e
+        log_error("#{e.message}. Sync ID: #{sync_run.sync_id}, Sync Run ID: #{sync_run.id}.")
+      end
+
+      def build_sync_records_in_parallel(records, sync_run, model)
+        Parallel.map(records, in_threads: THREAD_COUNT) do |message|
```
Comment (subintp) [inline lib/reverse_etl/extractors/full_refresh.rb:51]: Why do we need parellel here ? Building sync record is not touching any db or expensive operation.

File: spec/lib/reverse_etl/extractors/full_refresh_spec.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        subject.read(sync_run2.id, activity)
+        sync_run2.reload
+        sync_records = SyncRecord.where(sync_id: sync_run2.sync_id)
+        expect(sync_records.first.record["first_name"]).to eq("Mark")
+        expect(sync_records.count).to eq(1)
+        expect(sync_run2).to have_state(:queued)
+      end
+    end
+
+    context "with invalid satate" do
```
Comment (subintp) [inline spec/lib/reverse_etl/extractors/full_refresh_spec.rb:87]: typo

File: spec/lib/reverse_etl/loaders/standard_spec.rb [hunk 4]
```diff
@@ -166,5 +166,59 @@
         expect(sync_run_started).to have_state(:started)
       end
     end
+
+    context "full refresh clear record failed for individual" do
```
Comment (subintp) [inline spec/lib/reverse_etl/loaders/standard_spec.rb:170]: Context message is not clear.

File: spec/lib/reverse_etl/loaders/standard_spec.rb [hunk 5]
```diff
@@ -166,5 +166,59 @@
...
+          subject.write(sync_run_individual.id, activity)
+        end.to raise_error(Activities::LoaderActivity::FullRefreshFailed)
+
+        sync_run_individual.reload
+
+        expect(sync_run_individual).to have_state(:failed)
+      end
+    end
+
+    context "full refresh clear record failed for batch" do
```
Comment (subintp) [inline spec/lib/reverse_etl/loaders/standard_spec.rb:197]: Add a success case for full_refresh sync

File: lib/reverse_etl/extractors/full_refresh.rb (hunk 2 above)
Comment (afthabvp) [reply lib/reverse_etl/extractors/full_refresh.rb:51]: i will comment TODO for add benchmark the scenarios

File: lib/reverse_etl/extractors/full_refresh.rb (hunk 1 above)
Comment (afthabvp) [reply lib/reverse_etl/extractors/full_refresh.rb:18]: i will add comment TODO for one activity after write to clear all records associated with sync

File: spec/lib/reverse_etl/loaders/standard_spec.rb (hunk 5 above)
Comment (afthabvp) [reply spec/lib/reverse_etl/loaders/standard_spec.rb:197]: Success scenario: It always returns tracking_message and already covered in spec

**Review by subintp — APPROVED** [review]

**Pull request merged**